### ISSUE TITLE
PCS is not clear if SVE VG can be changed

### ISSUE THREAD

**Issue by a74nh** [ARM-software/abi-aa#260]
The **[Procedure Call Standard for the Arm® 64-bit Architecture (AArch64)](https://github.com/ARM-software/abi-aa/blob/2a70c42d62e9c3eb5887fa50b71257f20daca6f9/aapcs64/aapcs64.rst)** states:

> Like a short vector (see [Short Vectors](https://github.com/ARM-software/abi-aa/blob/2a70c42d62e9c3eb5887fa50b71257f20daca6f9/aapcs64/aapcs64.rst#short-vectors)), a scalable vector is a machine type that is composed of repeated instances of one fundamental integral or floating-point type. The number of bytes in the vector is always VG×8, where VG is a runtime value determined by the execution environment. VG is an even integer greater than or equal to 2; the ABI does not define an upper bound. VG is the same for all scalable vector types and scalable predicate types.

This states that VG is a runtime value. It is not explicit in stating if this is a _fixed_ value.

If a program whilst executing were to use Linux syscall `prctl()` to change the vector length, then that would change VG. It's not clear if this breaks the Procedure Call Standard. Is it ok for a function to change the vector length, call new functions, reset the vector length to it's previous value, then return? What happens if a function changes the vector length then returns without restoring the original value?

It would be useful if there was an additional paragraph clarifying this.

**Comment by yury-khrustalev** [comment]
The AAPCS64 says (6.1.3 Scalable vector registers) that

> If a subroutine takes at least one argument in scalable vector registers or scalable predicate registers, or returns results in such registers, the subroutine must ensure that **the entire contents of z8-z23 are preserved across the call**

There is a similar requirement for the contents of predicate registers.

What happens to the contents of vector registers when VL changes depends on the underlying target system. In case of Linux, as outlined in https://docs.kernel.org/arch/arm64/sve.html,

> Changing the vector length causes all of P0..P15, FFR and all bits of Z0..Z31 except for Z0 bits [127:0] .. Z31 bits [127:0] to become unspecified

so, if a function takes at least one argument in scalable vector registers or scalable predicate registers, or returns results in such registers, changing VL inside this function would break PCS rules unless extra care is taken to preserve the contents.

In some cases, provided that the care is taken (likely with inline assembly being involved), it's possible to change VL on the way and not break things, but this depends on the implementation and a compiler doesn't have to guarantee this.

I think adding more clarification to the AAPCS64 is not useful because we would have to refer to quirks of particular implementations, but it probably makes sense to add a clarification to the "Linux Implementation Notes" section in the SYSVABI64 document.

**Comment by AlfieRichardsArm** [comment]
I agree with you @yury-khrustalev.

I think it is also worth considering the impact of changing SVE VG if any vectors were on the stack, which could be relevant for passing through PCS.

As we can use VG for stack offsets, allowing the VG to change between a function call and a return would introduce a large amount of complexity for this.

Similarly, for stack unwinding, if the VG size was changes somewhere in the stack, this would introduce a tremendous amount of complexity to correctly unwind.

I think this was the TLDR of the discussion where we concluded the only time you could change VG safely was before entry in a program (regardless of OS/baremetal). But my memory is a bit hazy.

**Comment by Wilco1** [comment]
Yes. The address of a local is SP/FP + VL * C1 + C2. When C1 != 0, the computation fundamentally relies on VL never changing inside the function. And this applies to any local (eg. an address-taken integer).
